File: Problems & Solutions/1626. Best Team With No Conflicts/31 January 2023/Approach 3 - Binary Indexed Tree (BIT) - Fenwick Tree/Best Team With No Conflicts.java
class Solution {
    public int bestTeamScore(int[] scores, int[] ages) {
        int N = ages.length;
        int[][] ageScorePair = new int[N][2];

        for (int i = 0; i < N; i++) {
            ageScorePair[i][0] = scores[i];
            ageScorePair[i][1] = ages[i];
        }

        Arrays.sort(ageScorePair, (a,b) -> a[0] == b[0] ? a[1]-b[1] : a[0]-b[0]);

        int highestAge = 0;
        for (int i : ages) {
            highestAge = Math.max(highestAge, i);
        }
        int[] BIT = new int[highestAge + 1];

        int answer = Integer.MIN_VALUE;
        for (int[] ageScore : ageScorePair) {
            int currentBest = ageScore[0] + queryBIT(BIT, ageScore[1]);
            updateBIT(BIT, ageScore[1], currentBest);

            answer = Math.max(answer, currentBest);
        }

        return answer;
    }

    private int queryBIT(int[] BIT, int age) {
        int maxScore = Integer.MIN_VALUE;
        for (int i = age; i > 0; i -= i & (-i)) {
            maxScore = Math.max(maxScore, BIT[i]);
        }
        return maxScore;
    }

    private void updateBIT(int[] BIT, int age, int currentBest) {
        for (int i = age; i < BIT.length; i += i & (-i)) {
            BIT[i] = Math.max(BIT[i], currentBest);
        }
    }
}

File: Problems & Solutions/1626. Best Team With No Conflicts/31 January 2023/Approach 3 - Binary Indexed Tree (BIT) - Fenwick Tree/Best Team With No Conflicts.cpp
class Solution {
public:
    int bestTeamScore(vector<int>& scores, vector<int>& ages) {
        vector<pair<int, int>> ageScorePair;
        for (int i = 0; i < scores.size(); i++) {
            ageScorePair.push_back({scores[i], ages[i]});
        }
        sort(ageScorePair.begin(), ageScorePair.end());

        int highestAge = 0;
        for (int i : ages) {
            highestAge = max(highestAge, i);
        }
        vector<int> BIT(highestAge + 1, 0);
        
        int answer = INT_MIN;
        for (pair<int, int> ageScore : ageScorePair) {
            int currentBest = ageScore.first + queryBIT(BIT, ageScore.second);
            updateBIT(BIT, ageScore.second, currentBest);
            
            answer = max(answer, currentBest);
        }

        return answer;
    }
    
    int queryBIT(vector<int>& BIT, int age) {
        int maxScore = INT_MIN;
        for (int i = age; i > 0; i -= i & (-i)) {
            maxScore = max(maxScore, BIT[i]);
        }
        return maxScore;
    }


    void updateBIT(vector<int>& BIT, int age, int currentBest) {
        for (int i = age; i < BIT.size(); i += i & (-i)) {
            BIT[i] = max(BIT[i], currentBest);
        }
    }
};
